lsp/src/Language/LSP/Server/Core.hs

```diff
@@ -274,20 +300,36 @@ instance E.Exception ProgressCancelledException
 -- @since 0.11.0.0
 data ProgressCancellable = Cancellable | NotCancellable
 
+-- See Note [LSP configuration] for discussion of the configuration-related fields
 -- | Contains all the callbacks to use for initialized the language server.
 -- it is parameterized over a config type variable representing the type for the
 -- specific configuration data the language server needs to use.
 data ServerDefinition config = forall m a.
   ServerDefinition
     { defaultConfig :: config
       -- ^ The default value we initialize the config variable to.
-    , onConfigurationChange :: config -> J.Value -> Either T.Text config
-      -- ^ @onConfigurationChange oldConfig newConfig@ is called whenever the
-      -- clients sends a message with a changed client configuration. This
-      -- callback should return either the parsed configuration data or an error
-      -- indicating what went wrong. The parsed configuration object will be
-      -- stored internally and can be accessed via 'config'.
-      -- It is also called on the `initializationOptions` field of the InitializeParams
+    , configSection :: T.Text
```

Comment:
Is this always have the same value with `resConfigSection`?

Reply:
Yeah, we copy over a bunch of fields from `ServerDefinition` into `LanguageContextEnv`...